buildSrc/src/main/kotlin/CodegenTestCommon.kt

```diff
@@ -102,3 +107,135 @@ fun cargoCommands(properties: PropertyRetriever): List<Cargo> {
+                        rootProject.projectDir.absolutePath,
+                        pluginName,
+                        codegenTests(properties, allCodegenTests)
+                    )
+                )
+
+            // If this is a rebuild, cache all the hashes of the generated Rust files. These are later used by the
+            // `modifyMtime` task.
+            project.extra["previousBuildHashes"] = project.buildDir.walk()
+                .filter { it.name.endsWith(".rs") || it.name == "Cargo.toml" || it.name == "Cargo.lock" }
+                .map {
+                    HashUtils.getCheckSumFromFile(it) to it.lastModified()
+                }
+                .toMap()
+        }
+    }
+}
+
+fun registerGenerateCargoWorkspaceTask(
+    rootProject: Project,
+    project: Project,
+    pluginName: String,
+    allCodegenTests: List<CodegenTest>,
+    workingDirUnderBuildDir: String
+) {
+    val properties = PropertyRetriever(rootProject, project)
+    project.tasks.register("generateCargoWorkspace") {
+        description = "generate Cargo.toml workspace file"
+        doFirst {
+            project.buildDir.resolve("$workingDirUnderBuildDir/Cargo.toml")
+                .writeText(generateCargoWorkspace(pluginName, codegenTests(properties, allCodegenTests)))
+        }
+    }
+}
+
+fun registerGenerateCargoConfigTomlTask(
+    project: Project,
+    outputDir: File
+) {
+    project.tasks.register("generateCargoConfigToml") {
+        description = "generate `.cargo/config.toml`"
+        doFirst {
+            outputDir.resolve(".cargo").mkdir()
+            outputDir.resolve(".cargo/config.toml")
+                .writeText(
+                    """
+                    [build]
+                    rustflags = ["--deny", "warnings"]
+                    """.trimIndent()
+                )
+        }
+    }
+}
+
+fun registerModifyMtimeTask(
+    project: Project
+) {
+    // Cargo uses `mtime` (among other factors) to determine whether a compilation unit needs a rebuild. While developing,
+    // it is likely that only a small number of the generated crate files are modified across rebuilds. This task compares
+    // the hashes of the newly generated files with the (previously cached) old ones, and restores their `mtime`s if the
+    // hashes coincide.
+    // Debugging tip: it is useful to run with `CARGO_LOG=cargo::core::compiler::fingerprint=trace` to learn why Cargo
+    // determines a compilation unit needs a rebuild.
+    // For more information see https://github.com/awslabs/smithy-rs/issues/1412.
+    project.tasks.register("modifyMtime") {
+        description = "modify Rust files' `mtime` if the contents did not change"
+        dependsOn("generateSmithyBuild")
+
+        doFirst {
+            val previousBuildHashes: Map<String, Long> = project.extra["previousBuildHashes"] as Map<String, Long>
+
+            project.buildDir.walk()
+                .filter { it.name.endsWith(".rs") || it.name == "Cargo.toml" }
```

Comment:
Should probably pull out a function for the filter since these checks are different from the ones on line 135 (assuming that's not intentional).